### PR TITLE
Land Phase 1-3: state machine, racks, federation, audit, auth, tests

### LINKED ISSUE #28: Task: Bish / Bash / Bosh
### Goal

Uplift and implement baseline for a server based command line driven allocator

### Notes (Optional)

_No response_

### PR DESCRIPTION
**Pull request by Ulrond**
## Summary

This PR lands the cumulative Phase 1, Phase 2 and Phase 3 work that has been on `feature/28-task-bish-bash-bosh` since March, plus the comprehensive test suite that grew alongside it. The original PR title ("Fix code quality, deprecations, and codebase audit") undersold the scope — this contains the full backend, dashboard, auth, federation, audit, and test infrastructure for the allocator. Closes #28.

The branch was rebased onto current `develop` (no conflicts — develop's allocate-route fixes were already absorbed by the branch's refactor, verified before rebase) and the 31MB `bin/xts_allocator` PyInstaller artefact was scrubbed from history so the repo no longer carries it.

## What lands here

**Phase 1 — MVP**
- 6-state device state machine (`free / allocated / testing / resetting / maintenance / offline`) with transition validation (`state_machine.py`).
- Rack model + expanded Device model with platform, tags, target_id, owner, allocation_expiry, software_version, external_equipment.
- Advanced search endpoints (`POST /list_slots` filters, `POST /devices/search` free-text + filters).
- Duration-based allocation with automatic expiry checker (background task).
- `target_id` allocation strategy (slot-first, platform fallback).

**Phase 2 — backend features**
- JWT auth with role hierarchy (`auth.py`): `ROLE_ADMIN`, `ROLE_ENGINEER`, `ROLE_READONLY`. `@require_auth` decorator.
- Rate limiting (`rate_limiter.py`): sliding-window, per-user-email identifier, applied to write endpoints.
- Federation (master/slave): `POST /register`, `POST /heartbeat`, `GET /servers`, `GET /servers/<id>/devices`, `GET /devices/federated`.
- Health + Prometheus metrics: `GET /health`, `GET /metrics`.
- HTML dashboard at `GET /dashboard` (templates/dashboard.html).
- Audit trail (`audit_log.py`, `audit_log_routes.py`): JWT login/logout, allocation, deallocation, state change, device CRUD, unauthorized access — queryable via `GET /audit/logs` (admin-only).
- Structured logging framework (`logging_config.py`).
- Multi-environment config (`config.py`): SQLite and PostgreSQL.

**Phase 3 — test execution tracking**
- `TestExecution` model with heartbeat + duration tracking.
- `/start_test`, `/test_heartbeat`, `/end_test` route group.
- `AllocationHistory` audit trail (already in 1.x but expanded here).
- New `GET /device/<id>/box_status` compact polling endpoint (this PR) — lightweight payload (state, target_id, owner_email, active_test summary, heartbeat age) so RAFT/XTS pollers can call it frequently without paying the cost of `/list_slots`.

**.xts command surface**
- `allocator.xts` (2400+ lines): full CLI surface for `xts allocator …` commands (alloc, dealloc, borrow, search, register, active_test, end_test, plus admin and guide groups).
- Served via `app.static('/allocator.xts', './allocator.xts')` so `xts_core` can pull the command surface dynamically.

**Database**
- Alembic migration system (`migrations/`, `migrate.sh`).
- Test-database isolation (`bin/db-clean`, `bin/db-status`, `bin/db-switch`).
- PostgreSQL migration guide.

**Tests — 411 passing, 0 skipped, 0 failed**
- 21 test files covering every route file, state machine, concurrency, auth, rate limiting, federation, audit, export, usage, .xts validation.
- The previously-skipped 3 federation tests are now un-skipped using a real in-process `http.server` slave (not `AsyncMock`) — exercises the full master→slave serialization path.
- New `tests/test_route_input_validation.py` (31 cases): bad-input shapes across the six route files TODO.md flagged as lacking validation. Where a test failed because the endpoint crashed with 500 on malformed input, the route now validates at the boundary and returns 400 (`routes/device_routes.py`, `rack_routes.py`, `export_routes.py`, `usage_routes.py`, `federation_routes.py`, `audit_log_routes.py`).
- New `tests/test_xts_core_contract.py` (11 cases): locks the wire shape `xts_core` and other automation clients consume — `/allocator.xts` serving, allocate/dealloc response keys, `/device/<id>/box_status` payload shape. Any future allocator change that breaks the contract fails loudly here.

**Address Copilot review (49 inline comments)**
- All non-alembic unused imports removed via `ruff --select F401` (46 fixes across 30 files).
- Migration `op`/`sa` imports retained — Alembic convention, required for `--autogenerate`.
- Exception-handling cleanups: `app.py:166` cancel-error handling, `app.py:208` reordered so `PermissionError` handler is reachable, `rate_limiter.py:173` documented `except Exception: pass`, `format_allocator_output.py:51` bare except narrowed to `(ValueError, TypeError, AttributeError)`.
- Dead variable removals (`test/demo_pi_simple.py:172`, `tests/test_concurrency.py:99`).
- `bin/xts_allocator` (31.5 MB PyInstaller ELF, flagged by @mhughesacn) deleted and scrubbed from branch history. `.gitignore` updated.
- `auth_headers_readonly` fixture is now exercised by the new role-boundary validation tests.

**CodeQL alert resolved**
- `py/incomplete-url-substring-sanitization` alert on `tests/test_allocator_xts.py:247` was a false positive on a test assertion. Tightened the assertion to check the canonical `https://github.com/rdkcentral/xts_allocator_server` URL prefix instead of the bare substring `"github.com"` — strictly stronger check and no longer pattern-matches the lint rule.

## Follow-ups (NOT in this PR)

- Companion `xts_core` work tracked in a new issue on `rdkcentral/xts_core`: alias-from-URL caching, directory alias scanning, schema includes, pointing the bundled `XTSAllocatorClient` plugin at the allocator's `/allocator.xts` endpoint (the plugin currently targets older `/allocate` / `/deallocate` paths — captured in `XTS_ENHANCEMENT_PLAN.md` on this branch and locked by the new contract tests).
- Remaining `TODO.md` items: pagination on `/list_slots`, OpenAPI spec coverage, async DB sessions, terminology alignment (`tags`/`labels`, `slot_contents`/`external_equipment`).

## Test plan

- [x] `pytest tests/ -v` — 411 passed / 0 skipped / 0 failed locally on the rebased branch.
- [x] All four GitHub Actions checks (Analyze (actions), Analyze (python), Signature/Check, copyright) green on a previous run; expecting same after force-push.
- [x] CodeQL alert on `tests/test_allocator_xts.py:247` resolved by tightening the assertion.
- [ ] Manual smoke: boot allocator standalone (`./run.sh`), `curl /health`, `curl /allocate_slot`, then drive same endpoints via `xts allocate --server …` to confirm both independent and xts-triggered flows work against the same instance.
- [ ] Reviewer sign-off from @mhughesacn (or current maintainer).

### CODE REVIEW

File: bin/xts_allocator
Comment (mhughesacn) [inline bin/xts_allocator:1]: This is a binary which should not be checked into git.  Please can it be deleted?

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    """Generate Authorization header for admin role."""
+    token = generate_token("admin@example.com", ROLE_ADMIN)
+    return {"Authorization": f"Bearer {token}"}
+
+
+@pytest.fixture
+def auth_headers_readonly():
+    """Generate Authorization header for readonly role."""
+    token = generate_token("viewer@example.com", ROLE_READONLY)
+    return {"Authorization": f"Bearer {token}"}
```
[the remark this answers is not in the document]
Comment (Ulrond) [reply tests/conftest.py:93]: Addressed in the latest push — `auth_headers_readonly` is now exercised by the new `tests/test_route_input_validation.py` (`TestDeviceRoutesInputValidation::test_add_slot_readonly_role_forbidden` and `TestAuditLogRoutesInputValidation::test_audit_logs_requires_admin_readonly_denied`) which assert that a readonly user gets 401/403 on write endpoints. The fixture is no longer defined-but-unused.

File: routes/utils.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+def ensure_utc(dt):
+    """Make a datetime timezone-aware (UTC) if it is naive.
+
+    SQLite strips timezone info from stored datetimes, so values read back
+    are naive.  This helper re-attaches UTC so they can be compared with
+    timezone-aware ``datetime.now(timezone.utc)`` values.
+    """
+    if dt is not None and dt.tzinfo is None:
+        return dt.replace(tzinfo=timezone.utc)
+    return dt
```
[the remark this answers is not in the document]
Comment (Ulrond) [reply routes/utils.py:15]: Deferring this codebase-wide datetime audit as a follow-up rather than expanding the scope of an already-large PR. The `ensure_utc()` helper itself stays in this PR and is used wherever SQLite returns naive datetimes (federation_routes, audit_log_routes, test_routes); the comprehensive audit would touch many more sites and is better as its own focused PR. I'll open a follow-up issue on this repo for the audit.

**Comment by Ulrond** [comment]
## Update — rebase + Phase 1-3 land, plus standalone CLI added

Force-pushed the rebased branch (40 commits) plus follow-up work. Current state:

**What's in HEAD now (e2a6c7e → 9b67294):**

1. Rebased onto `origin/develop` — git auto-detected that PR #23's three commits (and 6 others) were already absorbed by the branch's refactor, no conflicts. PR #23's `/allocate_slot` 404 / 409 / 200-empty-slots behaviours verified preserved before rebase.
2. `bin/xts_allocator` (31.5 MB PyInstaller artefact flagged by @mhughesacn) scrubbed from **all of branch history** via `git filter-branch`, not just deleted at HEAD. Repo no longer carries the binary in any commit.
3. 46 unused imports removed via `ruff --select F401 --fix` across 30 files (alembic `op`/`sa` retained per convention).
4. Exception-handling fixes from the Copilot review: `app.py:208` `PermissionError`/`OSError` reordered so the permission handler is reachable, `rate_limiter.py:173` annotated, `format_allocator_output.py:51` bare except narrowed.
5. 3 previously-skipped federation tests now un-skipped — use a real `http.server.HTTPServer` slave on an ephemeral port instead of the brittle `AsyncMock` approach the original author flagged. Exercises the full master→slave serialization path including the ECONNREFUSED branch (bind-then-close a socket to grab a dead port).
6. New `tests/test_route_input_validation.py` (31 cases) for the six route files TODO.md flagged as lacking validation. Where a route 500'd on malformed input, added a small `validate_string`/`_cap_query_param` guard at the boundary so it returns 400 instead. Also exercises the previously-unused `auth_headers_readonly` fixture.
7. New `tests/test_xts_core_contract.py` (11 cases) locking the wire shape automation clients consume — `/allocator.xts` static serving, `/allocate_slot` response keys (`slot_id`, `rackName`, `slotName`, `target_id`, `state`, `owner_email`, `allocation_history_id`), the new `/device/<id>/box_status` payload, and the 404 / 200-empty-slots branches.
8. New `GET /device/<id>/box_status` compact polling endpoint (TODO.md "compact box status endpoint for RAFT/XTS polling"). Lightweight payload only — `state`, `target_id`, `owner_email`, `allocation_expiry`, `active_test` with elapsed minutes and heartbeat age.
9. **New `bin/allocator` CLI** (~250 LOC, stdlib only — `argparse` + `urllib`). Drives the entire allocator REST API without xts/xts_core: `health`, `list-slots`, `add-slot`, `allocate`, `dealloc`, `box-status`, `start-test`, `heartbeat`, `end-test`, `smoke`, `xts`. The `smoke` subcommand walks the full lifecycle end-to-end (login → add-slot → allocate → start-test → heartbeat → box-status → end-test → dealloc) and exits non-zero on the first failure. Verified locally: 11/11 green against a freshly-booted instance. This is the canonical "trigger the allocator and run tests against it without xts plumbing" entry point.

**Tests: 411 passed / 0 skipped / 0 failed locally** (was 366 / 3 / 0 before this push).

**CI status as of this push:**
- ✅ copyright (Jenkins): pass
- ❌ blackduck (Jenkins): fail — Jenkins URL is on internal infra so I can't see the specific finding. Pre-existing or new from one of the deps added on the branch (`PyJWT`, `psycopg2-binary`, `sanic-testing`, `httpx`, `pyinstaller`, `alembic`)? Maintainer log access needed.
- ⏳ GitHub Actions (CodeQL Analyze, CLA Signature): not re-triggered yet — last runs were Feb 12/16 on `a736de6`. This appears to be the first-time-contributor approval gate; pinging maintainer to approve workflow runs on the rebased HEAD.

**Follow-ups filed:**
- [rdkcentral/xts_allocator_server#32](https://github.com/rdkcentral/xts_allocator_server/issues/32) — codebase-wide `ensure_utc()` audit (deferred from this PR's scope per Copilot recommendation; reply on the original Copilot comment).
- Comment on [rdkcentral/xts_core#59](https://github.com/rdkcentral/xts_core/issues/59) listing the xts_core specifics needed alongside this (alias-from-URL caching, schema `includes:`, retargeting the bundled `XTSAllocatorClient` plugin to current endpoint names — `/allocate_slot` / `/deallocate_slot`, not `/allocate` / `/deallocate`).

Ready for re-review. /cc @mhughesacn — would appreciate (a) approval for the GitHub Actions workflows to run on this rebased HEAD, and (b) eyes on the blackduck failure if you can pull the Jenkins log.

**Review by mhughesacn — COMMENTED** [review]
Sorry Gerald: I typed this up yesterday (Tuesday) but didn't submit properly.

Hi @Ulrond : 
* The files below are showing code matches.  The matches are not showing as clearly to an alembic component in Blackduck, but all appear to be related to the use of alembic.  e.g. Some matches are to other components but with "alembic" in the path.  Where did these files come from?

migrations/versions/732b0ff2af4c_initial_schema.py
migrations/versions/755c75b1348e_add_testexecution_table_and_complete_.py
migrations/env.py
migrations/script.py.mako

* All new significant code files should have headers (>maybe 10 lines).
